test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/DotnetCommand.cs

```diff
@@ -15,6 +16,35 @@ public DotnetCommand()
         public DotnetCommand(string dotnetUnderTest)
             : base(dotnetUnderTest)
         {
+            // Some scripts (e.g. RunCsc/RunVbc) depend on the correct `dotnet`
+            // being in the PATH. When tests use `DotnetUnderTest.WithBackwardsCompatibleRuntimes`,
+            // this means that the up-to-date dotnet command is resolved by the
+            // old-runtime-targeting RunCsc/Roslyn, causing compatability issues.
+            // So, detect if we're running a dotnet that's not just a plain "dotnet",
+            // and if so, make sure it's on the PATH.
+            var containingDirectory = Path.GetDirectoryName(dotnetUnderTest);
+            if (!string.IsNullOrEmpty(containingDirectory))
+            {
+                // if it's already set, use that
+                string path;
+                if (!Environment.TryGetValue("PATH", out path))
+                {
+                    // otherwise grab it from the environment
+                    path = System.Environment.GetEnvironmentVariable("PATH");
+                }
+                // if it's the only item in the path, set it
+                if (string.IsNullOrEmpty(path))
+                {
+                    path = containingDirectory;
+                }
+                else
+                {
+                    // otherwise, join it with the rest of the path,
+                    // putting it in front so it takes priority
+                    path = containingDirectory + Path.PathSeparator + path;
```

Comment:
Nothing harmful - the directory will just be searched twice when looking up an exe. (I just pushed a commit that solves this in the basic case, but ignores all the complex cases that filesystems can introduce, like symlinks)